numpy/core/src/multiarray/descriptor.c

```diff
@@ -3754,69 +3754,91 @@ descr_repeat(PyObject *self, Py_ssize_t length)
+    if (PyDataType_HASFIELDS(self)) return (PyObject *) self;
+    PyObject *astr = arraydescr_str(self);
 #if defined(NPY_PY3K)
-        PyObject *bstr = PyUnicode_AsUnicodeEscapeString(astr);
-        Py_DECREF(astr);
-        astr = bstr;
+    PyObject *bstr = PyUnicode_AsUnicodeEscapeString(astr);
+    Py_DECREF(astr);
+    astr = bstr;
 #endif
-        PyErr_Format(PyExc_KeyError,
+    PyErr_Format(PyExc_KeyError,
                 "There are no fields in dtype %s.", PyBytes_AsString(astr));
-        Py_DECREF(astr);
+    Py_DECREF(astr);
+    return NULL;
+}
+
+static PyObject *
+_subscript_by_name(PyArray_Descr *self, PyObject *op)
+{
+    PyObject *obj = PyDict_GetItem(self->fields, op);
+    PyObject *descr;
+    PyObject *s;
+
+    if (obj == NULL) {
+        if (PyUnicode_Check(op)) {
+            s = PyUnicode_AsUnicodeEscapeString(op);
+        }
+        else {
+            s = op;
+        }
+
+        PyErr_Format(PyExc_KeyError,
+                "Field named \'%s\' not found.", PyBytes_AsString(s));
+        if (s != op) {
+            Py_DECREF(s);
+        }
+        return NULL;
+    }
+    descr = PyTuple_GET_ITEM(obj, 0);
+    Py_INCREF(descr);
+    return descr;
+}
+
+static PyObject *
+_subscript_by_index(PyArray_Descr *self, Py_ssize_t i)
+{
+    PyObject *name = PySequence_GetItem(self->names, i);
+    if(name == NULL) {
+        PyErr_Format(PyExc_IndexError,
+                     "Field index %d out of range.", i);
+        return NULL;
+    }
+    return _subscript_by_name(self, name);
+}
+
+static PyObject *
+descr_item(PyArray_Descr *self, Py_ssize_t i)
+{
+    PyObject *name;
+    if(_check_has_fields(self) == NULL) {
         return NULL;
     }
+    return _subscript_by_index(self, i);
+}
+
+static PyObject *
+descr_subscript(PyArray_Descr *self, PyObject *op)
+{
+    PyObject *retval;
+    if(_check_has_fields(self) == NULL) {
+        return NULL;
+    }
+
 #if defined(NPY_PY3K)
     if (PyUString_Check(op)) {
 #else
     if (PyUString_Check(op) || PyUnicode_Check(op)) {
 #endif
-        PyObject *obj = PyDict_GetItem(self->fields, op);
-        PyObject *descr;
-        PyObject *s;
-
-        if (obj == NULL) {
-            if (PyUnicode_Check(op)) {
-                s = PyUnicode_AsUnicodeEscapeString(op);
-            }
-            else {
-                s = op;
-            }
-
-            PyErr_Format(PyExc_KeyError,
-                    "Field named \'%s\' not found.", PyBytes_AsString(s));
-            if (s != op) {
-                Py_DECREF(s);
-            }
-            return NULL;
-        }
-        descr = PyTuple_GET_ITEM(obj, 0);
-        Py_INCREF(descr);
-        retval = descr;
+        return _subscript_by_name(self, op);
     }
     else if (PyInt_Check(op)) {
-        PyObject *name;
-        int size = PyTuple_GET_SIZE(self->names);
-        int value = PyArray_PyIntAsInt(op);
-        int orig_value = value;
-
+        Py_ssize_t i = PyInt_AsSsize_t(op);
```

Comment:
Which bit? I agree that despite being non-guaranteed, it would be alarming if that turned out to be an issue